src/supported-operations/thin-edge-supported-operations.md

```diff
@@ -0,0 +1,118 @@
+  # Required
+  command = "echo"
+  # Optional
+  root = true
+
+[mqtt]
+  # MQTT configuration if the operation requires MQTT communication, e.g. forwarding message JSON on the bus
+  topic = "tedge/logs"
+
+[extras]
+  # Additional configuration if the operation requires additional configuration
+  log_type = ["error"]
+```
+
+* If the operation doesn't require additional configuration, an empty `toml` file can be used.
+* Basic tables names are fixed and are `exec`, `mqtt` and `extras`. Additional tables can be added as needed.
+* Tables `[exec]` or `[mqtt]` must be present if config is not empty.
+* Tables `[exec]` and `[mqtt]` are mutually exclusive and only one of them can be used.
+
+## Example 1
+
+Given operation `restart`, the operation file would be created in `/etc/tegde/operations/c8y` with the filename `c8y_Restart` so the full path would be `/etc/tegde/operations/c8y/c8y_Restart`.
+Assuming `restart` operation requires no additional parameters as well as can be executed directly in the executor (assume a mapper) no additional configuration is required.
+
+An empty `toml` file is still valid `toml` file and can be used to indicate no additional configuration is required.
+
+Given a mapper the flow would be as follows:
+
+1. Mapper reads directory for the cloud which it supports i.e. `/etc/tegde/operations/c8y`.
+2. Mapper finds the operation file for the operation `restart` in the directory `c8y` with filename `c8y_Restart` which is the only operation.
+3. Mapper takes the filename as operation name and reads the operation file.
+4. Empty operation file means there are no additional configuration parameters and mapper is ready to send supported operations message to c8y which contains the list of supported operations i.e. `c8y_Restart`.
+5. Cloud operator wants restart the device therefore they send the operation message to the device which mapper interprets as a restart operation and mapper executes restart.
+
+## Example 2 (special case)
+
+Given operation `c8y_LogfileRequest`, the operation file would be created in `/etc/tegde/operations/c8y` with the filename `c8y_LogfileRequest` so the full path would be `/etc/tegde/operations/c8y/c8y_LogfileRequest`.
+Operation `c8y_LogfileRequest` requires additional parameter `log_type` to be set.
+
+With the operation file the following structure is written:
+
+```toml
+[exec]
+
+command = "/etc/tedge/plugins/c8y_LogfileRequest"
+user = "root"
+
+[init]
+topic = "c8y/s/us"
+message = "116,error"
+
+[extras]
+  # Additional configuration if the operation requires additional configuration
+  log_type = ["error"]
+```
+
+> Alternatives:
+>
+> Option 1: on init of mapper config will contain a command to be called e.g. sent `log_type` message to c8y, only the component knows about it (mapper doesn't need to care about it)
+> Option 2: explicit additional message to be send by the mapper if required/defined using a table in config file e.g. `[init]` like in the example above.
```

Comment:
My preference would be to avoid the plugins to be forced to send MQTT messages on their own, especially in the on-demand executable mode. If we have the requirement for the operation plugin to send more complex/dynamic messages on initialisation, I would even make it print the message payload to the stdout which tedge can read and send to the MQTT broker on its behalf.

On the other hand, for an operation like `c8y_LogfileRequest` which even needs to send HTTP messages to upload the log file etc, sending an extra MQTT message wouldn't be a big deal. That would simplify the c8y-mapper code quite a lot. If we go down that path, we could even make the plugin report the success or failure of the operation directly via SmartREST messages published to `c8y/s/us`.

Reply:
On the reverse, I prefer to let all the operation specific protocol responsibility to the operation plugin.

* The purpose of having operation plugins is to be *open* to new operations provided *independently*.
   * The protocol for an operation is designed by Cumulocity and might be very different from an operation to another.  
   * One cannot model in advance all these specificities.
* The support brings by thin-edge to these plugin must focus on the *coordination* of the operations
   * Without any support, a contributor can already implement an operation on a thin-edge device, *but* these contributions fight together to declare their capabilities to Cumulocity.
   * Awaking a plugin only when required, i.e. when an operation has been triggered, is the second level of support thin-edge can bring with no knowledge of the protocol (beyond the triggering event).
   * Making sure the operations are executed in concert and reliably is the third level of support. This point is the where thin-edge can add true benefits. This is also the point that still needs the more clarifications.